src/WebJobs.Script.WebHost/DependencyInjection/DependencyValidator/ServiceMatch.cs

```diff
@@ -0,0 +1,202 @@
+        }
+
+        public void Add<TImplementation>(ServiceLifetime lifetime)
+        {
+            ServiceDescriptor desc = new ServiceDescriptor(ServiceType, typeof(TImplementation), lifetime);
+            _requiredDescriptors.Add(desc);
+        }
+
+        public void Add<TPeerType>(string typeName, ServiceLifetime lifetime)
+        {
+            if (typeName == null)
+            {
+                throw new ArgumentNullException(nameof(typeName));
+            }
+
+            Type peerType = typeof(TPeerType);
+            Add(peerType, typeName, lifetime);
+        }
+
+        public void Add(string typeName, ServiceLifetime lifetime)
+        {
+            if (typeName == null)
+            {
+                throw new ArgumentNullException(nameof(typeName));
+            }
+
+            Add(ServiceType, typeName, lifetime);
+        }
+
+        private void Add(Type peerType, string typeName, ServiceLifetime lifetime)
+        {
+            // So we don't have to build the full type each time we rev an assembly,
+            // use a peer type to construct the full name
+            string fullTypeName = peerType.AssemblyQualifiedName.Replace(peerType.FullName, typeName);
+
+            ServiceDescriptor desc = new ServiceDescriptor(ServiceType, Type.GetType(fullTypeName), lifetime);
+            _requiredDescriptors.Add(desc);
+        }
+
+        public void AddFactory<TFactoryType>(ServiceLifetime lifetime)
+        {
+            // Hijack the factory to pass in the expected type
+            ServiceDescriptor desc = new ServiceDescriptor(ServiceType, s => typeof(TFactoryType).Assembly, lifetime);
+            _requiredDescriptors.Add(desc);
+        }
+
+        public void AddOptional<TImplementation>(ServiceLifetime lifetime)
+        {
+            if (_match != MatchType.Collection)
+            {
+                throw new InvalidOperationException("Optional matches only apply to Collections.");
+            }
+
+            ServiceDescriptor desc = new ServiceDescriptor(ServiceType, typeof(TImplementation), lifetime);
+            _optionalDescriptors.Add(desc);
+        }
+
+        public IEnumerable<InvalidServiceDescriptor> FindInvalidServices(IServiceCollection services)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            IEnumerable<ServiceDescriptor> registered = services.Where(p => p.ServiceType == ServiceType);
+
+            switch (_match)
+            {
+                case MatchType.None:
+                    // If we expect none, all of the registered services for this type are invalid
+                    return registered.Select(p => new InvalidServiceDescriptor(p, InvalidServiceDescriptorReason.Invalid));
+
+                case MatchType.Single:
+                    ServiceDescriptor expected = _requiredDescriptors.Single();
```

Comment:
Do we have a test validating this? If the expectation is single, but that expectation is not met, this call will throw

Reply:
If you try to use the builder API, it'll block you from adding more than one. I've just added a test for that specifically. And the API ensures that you have at least one, so having zero wouldn't happen either.

However, even if we do throw, it'd be okay and show you that you built your expectations wrong somehow. That .Single() call is on the "expected" collection -- no registered services could alter that collection. In other words, it'd be a dev-time failure.